File: caffe_3d/action_matlab/VideoSpatialPrediction.m
function prediction = VideoSpatialPrediction(vid_name, mean_file, net)
num=25;
vidObj = VideoReader(vid_name);
duration = vidObj.NumberOfFrames;
step = floor((duration-1)/(num-1));
rgb = zeros(256,340,3,num,'single');
rgb_flip = zeros(256,340,3,num,'single');


for i = 1:num
    img = single(imresize(read(vidObj,(i-1)*step+1),[256,340]));
    rgb(:,:,:,i) = img;
    rgb_flip(:,:,:,i) = img(:,end:-1:1,:);
end

rgb_1 = rgb(1:224,1:224,:,:);
rgb_2 = rgb(1:224,end-223:end,:,:);
rgb_3 = rgb(16:16+223,60:60+223,:,:);
rgb_4 = rgb(end-223:end,1:224,:,:);
rgb_5 = rgb(end-223:end,end-223:end,:,:);
rgb_f_1 = rgb_flip(1:224,1:224,:,:);
rgb_f_2 = rgb_flip(1:224,end-223:end,:,:);
rgb_f_3 = rgb_flip(16:16+223,60:60+223,:,:);
rgb_f_4 = rgb_flip(end-223:end,1:224,:,:);
rgb_f_5 = rgb_flip(end-223:end,end-223:end,:,:);

rgb = cat(4,rgb_1,rgb_2,rgb_3,rgb_4,rgb_5,rgb_f_1,rgb_f_2,rgb_f_3,rgb_f_4,rgb_f_5);
d = load(mean_file);
IMAGE_MEAN = single(d.image_mean);
rgb = bsxfun(@minus,rgb(:,:,[3,2,1],:),IMAGE_MEAN);
rgb = permute(rgb,[2,1,3,4]);



prediction = zeros(101,size(rgb,4));
batch_size = 50;
num_batches = ceil(size(rgb,4)/batch_size);
rgbs = zeros(224,224,3,batch_size,'single');

for bb = 1:num_batches
	range = 1 + batch_size*(bb-1): min(size(rgb,4),batch_size*bb);
	rgbs(:,:,:,mod(range-1,batch_size)+1) = single(rgb(:,:,:,range));
	out_put = net.forward({rgbs});
	out_put = squeeze(out_put{1});
	prediction(:,range) = out_put(:,mod(range-1,batch_size)+1);
end

end

File: caffe_3d/action_python/VideoSpatialPrediction.py
'''
A sample function for classification using spatial network
Customize as needed:
e.g. num_categories, layer for feature extraction, batch_size
'''

import glob
import os
import numpy as np
import math
import cv2
import scipy.io as sio

def VideoSpatialPrediction(
        vid_name,
        mean_file,
        net,
        num_categories,
        feature_layer,
        start_frame=0,
        num_frames=0,
        num_samples=25
        ):

    if num_frames == 0:
        imglist = glob.glob(os.path.join(vid_name, '*image_*.jpg'))
        duration = len(imglist)
    else:
        duration = num_frames

    # selection
    step = int(math.floor((duration-1)/(num_samples-1)))
    dims = (256,340,3,num_samples)
    rgb = np.zeros(shape=dims, dtype=np.float64)
    rgb_flip = np.zeros(shape=dims, dtype=np.float64)

    for i in range(num_samples):
        img_file = os.path.join(vid_name, 'image_{0:05d}.jpg'.format(i*step+1))
        img = cv2.imread(img_file, cv2.IMREAD_UNCHANGED)
        img = cv2.resize(img, dims[1::-1])
        rgb[:,:,:,i] = img
        rgb_flip[:,:,:,i] = img[:,::-1,:]

    # crop
    rgb_1 = rgb[:224, :224, :,:]
    rgb_2 = rgb[:224, -224:, :,:]
    rgb_3 = rgb[16:240, 60:284, :,:]
    rgb_4 = rgb[-224:, :224, :,:]
    rgb_5 = rgb[-224:, -224:, :,:]
    rgb_f_1 = rgb_flip[:224, :224, :,:]
    rgb_f_2 = rgb_flip[:224, -224:, :,:]
    rgb_f_3 = rgb_flip[16:240, 60:284, :,:]
    rgb_f_4 = rgb_flip[-224:, :224, :,:]
    rgb_f_5 = rgb_flip[-224:, -224:, :,:]

    rgb = np.concatenate((rgb_1,rgb_2,rgb_3,rgb_4,rgb_5,rgb_f_1,rgb_f_2,rgb_f_3,rgb_f_4,rgb_f_5), axis=3)

    # substract mean
    d = sio.loadmat(mean_file)
    image_mean = d['image_mean']

    rgb = rgb[...] - np.tile(image_mean[...,np.newaxis], (1, 1, 1, rgb.shape[3]))
    rgb = np.transpose(rgb, (1,0,2,3))

    # test
    batch_size = 50
    prediction = np.zeros((num_categories,rgb.shape[3]))
    num_batches = int(math.ceil(float(rgb.shape[3])/batch_size))

    for bb in range(num_batches):
        span = range(batch_size*bb, min(rgb.shape[3],batch_size*(bb+1)))
        net.blobs['data'].data[...] = np.transpose(rgb[:,:,:,span], (3,2,1,0))
        output = net.forward()
        prediction[:, span] = np.transpose(output[feature_layer])

    return prediction
